lib/datadog/appsec/event.rb

```diff
@@ -34,86 +36,120 @@ module Event
+              tags['network.client.ip'] = request.remote_addr
             end
 
-            request_headers.each do |header, value|
-              tags["http.request.headers.#{header}"] = value
+            if (response = event[:response])
+              response.headers.each do |header, value|
+                tags["http.response.headers.#{header}"] = value if ALLOWED_RESPONSE_HEADERS.include?(header.downcase)
+              end
             end
 
-            tags['http.host'] = request.host
-            tags['http.useragent'] = request.user_agent
-            tags['network.client.ip'] = request.remote_addr
-          end
+            waf_result = event[:waf_result]
+            # accumulate triggers
+            waf_events += waf_result.events
 
-          if (response = event[:response])
-            response_headers = response.headers.select do |k, _|
-              ALLOWED_RESPONSE_HEADERS.include?(k.downcase)
-            end
+            waf_result.derivatives.each do |key, value|
+              parsed_value = json_parse(value)
+              next unless parsed_value
 
-            response_headers.each do |header, value|
-              tags["http.response.headers.#{header}"] = value
+              parsed_value_size = parsed_value.size
+
+              compressed_data = compressed_and_base64_encoded(parsed_value)
+              compressed_data_size = compressed_data.size
+
+              if compressed_data_size >= MAX_ENCODED_SCHEMA_SIZE && parsed_value_size >= MAX_ENCODED_SCHEMA_SIZE
+                Datadog.logger.debug do
+                  "Schema key: #{key} exceeds the max size value. It will not be included as part of the span tags"
+                end
+                next
+              end
+
+              derivative_value = parsed_value_size > compressed_data_size ? compressed_data : parsed_value
+
+              tags[key] = derivative_value
             end
+
+            tags
           end
 
-          tags['_dd.origin'] = 'appsec'
+          appsec_events = json_parse({ triggers: waf_events })
+          entry_tags['_dd.appsec.json'] = appsec_events if appsec_events
+          entry_tags
+        end
+        # rubocop: enable Metrics/MethodLength
 
-          # accumulate triggers
-          waf_result = event[:waf_result]
-          tags['_dd.appsec.triggers'] ||= []
-          tags['_dd.appsec.triggers'] += waf_result.events
+        private
 
-          waf_result.derivatives.each do |key, value|
-            tags[key] = JSON.dump(value)
-          end
+        def compressed_and_base64_encoded(value)
+          Base64.encode64(gzip(value))
+        rescue TypeError
+          nil
+        end
+
+        def json_parse(value)
+          JSON.dump(value)
+        rescue ArgumentError
+          nil
+        end
 
-          tags
+        def gzip(value)
+          sio = StringIO.new
+          gz = Zlib::GzipWriter.new(sio, Zlib::DEFAULT_COMPRESSION, Zlib::DEFAULT_STRATEGY)
```

Comment:
Could I trouble to run very simple benchmarks in your machine with a sample `value` payload and different combinations of compression and strategy arguments?